File: js/test/test.BBBrowserExtensionInjector.ts
import assert from "node:assert/strict";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, it } from "vitest";

import { ModCDPClient } from "../src/client/ModCDPClient.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "..", "dist", "extension");

describe("BBBrowserExtensionInjector", () => {
  it(
    "uploads the real extension and launches a Browserbase browser with it installed",
    { timeout: 180_000 },
    async () => {
      assert.ok(process.env.BROWSERBASE_API_KEY?.trim(), "BROWSERBASE_API_KEY is required for live Browserbase tests");
      const cdp = new ModCDPClient({
        launcher: {
          launcher_mode: "bb",
          launcher_options: {
            timeout: 120,
            ...(process.env.BROWSERBASE_REGION ? { region: process.env.BROWSERBASE_REGION } : {}),
          },
        },
        upstream: { upstream_mode: "ws" },
        injector: {
          injector_mode: "inject",
          injector_extension_path: EXTENSION_PATH,
          injector_service_worker_url_suffixes: ["/modcdp/service_worker.js"],
          injector_trust_service_worker_target: true,
        },
      });

      try {
        await cdp.connect();
        assert.equal(cdp.connect_timing?.injector_source, "bb");
        assert.equal(typeof cdp.extension_id, "string");
        const service_worker_url = await cdp.Mod.evaluate({
          expression: "chrome.runtime.getURL('modcdp/service_worker.js')",
        });
        assert.match(String(service_worker_url), /^chrome-extension:\/\/[a-z]{32}\/modcdp\/service_worker\.js$/);
      } finally {
        await cdp.close();
      }
    },
  );
});

File: python/tests/test_BBBrowserExtensionInjector.py
from __future__ import annotations

import os
from pathlib import Path
import unittest

from modcdp import ModCDPClient

HERE = Path(__file__).resolve().parent
EXTENSION_PATH = HERE.parents[1] / "dist" / "extension"


class BBBrowserExtensionInjectorTests(unittest.TestCase):
    def test_uploads_real_extension_and_launches_browserbase_browser_with_it_installed(self) -> None:
        if not os.environ.get("BROWSERBASE_API_KEY", "").strip():
            self.fail("BROWSERBASE_API_KEY is required for live Browserbase tests")
        cdp = ModCDPClient(
            launcher={
                "launcher_mode": "bb",
                "launcher_options": {
                    "timeout": 120,
                    **({"region": os.environ["BROWSERBASE_REGION"]} if os.environ.get("BROWSERBASE_REGION") else {}),
                },
            },
            upstream={"upstream_mode": "ws"},
            injector={
                "injector_mode": "inject",
                "injector_extension_path": str(EXTENSION_PATH),
                "injector_service_worker_url_suffixes": ["/modcdp/service_worker.js"],
                "injector_trust_service_worker_target": True,
            },
        )

        try:
            cdp.connect()
            self.assertEqual(cdp.connect_timing.get("injector_source") if cdp.connect_timing else None, "bb")
            self.assertIsInstance(cdp.extension_id, str)
            service_worker_url = cdp.Mod.evaluate(expression="chrome.runtime.getURL('modcdp/service_worker.js')")
            self.assertRegex(str(service_worker_url), r"^chrome-extension://[a-z]{32}/modcdp/service_worker\.js$")
        finally:
            cdp.close()


if __name__ == "__main__":
    unittest.main()
